### PR TITLE
Resolve syncRecords

### PR DESCRIPTION
**Pull request by ayumi**

### CODE REVIEW

File: client/recordUtil.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  /**
+   * In case of UPDATEs to nonexistant records, try falling back to CREATE.
+   * @param {SyncRecord} record
+   * @returns {SyncRecord=} SyncRecord with action CREATE; or null.
+   */
+  static createFromUpdate (record) {
+    if (!record || record.action !== proto.actions.UPDATE) {
+      throw new Error('Missing UPDATE syncRecord.')
+    }
+    switch (record.objectData) {
```
Comment (diracdeltas) [inline client/recordUtil.js]: seems ok for these all to just return record.objectData + record.objectId

File: client/recordUtil.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    const site = record.historySite
+    const newSite = {
+      location: site.location,
+      title: site.title || site.location,
+      customTitle: site.customTitle || undefined,
+      lastAccessedTime: site.lastAccessedTime || Date.now(),
+      creationTime: site.creationTime || Date.now()
+    }
+    return record.$type.create(Object.assign(
```
Comment (ayumi) [inline client/recordUtil.js]: to confirm: a protobuf object is created here but not encoded into bytes, so if it's sent to the browser via IPC will it be able to access the object?

File: client/recordUtil.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * You have a collection of records which are synced.
+   * Given a new SyncRecord and the matching existing data record (if available),
+   * this function resolves the action on your existing record collection.
+   * Browser clients periodically request records, then finds matching browser
+   * objects and IPC calls this function to get resolved browser writes.
+   * @param {SyncRecord} record
+   * @param {Object=} existingObject (e.g. { location: "https://...", title: "cool", ... })
+   * @returns {SyncRecord} resolved record which the browser should do
```
Comment (ayumi) [inline client/recordUtil.js]: I think that makes sense when the browser provides an existingObject.
Without an existingObject, for example a Create record, Sync would still need to send the SyncRecord to the browser for the browser to create its object.
Do we want `resolve()` to return multiple "types": browser object w/ merged SyncRecord, original SyncRecord or null?

**Comment by ayumi** [comment]
@diracdeltas i've made changes to reflect resolution of syncRecords + existingObjects. could you take another look?

File: client/constants/messages.js [hunk 4]
```diff
@@ -45,12 +45,24 @@ const messages = {
   FETCH_SYNC_RECORDS: _, /* @param Array.<string> categoryNames, @param {number} startAt (in seconds) */
   /**
    * webview -> browser
-   * browser must update its local values with the new sync records, performing
-   * conflict-resolution as necessary.
+   * browser should find objects with matching objectIds then call
+   * RESOLVE_SYNC_RECORDS to perform conflict resolution.
    */
   RECEIVE_SYNC_RECORDS: _, /* @param {string} categoryName, @param {Array.<Object>} records */
```
Comment (diracdeltas) [inline client/constants/messages.js]: i wonder if categoryName is needed any more; browser can infer category from the record fields?

Comment (diracdeltas) [inline client/constants/messages.js]: also this should be renamed to something like `GET_EXISTING_RECORDS`

File: package.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
     "flow": "flow; test $? -eq 0 -o $? -eq 2",
     "build": "yarn upgrade && browserify client/sync.js | uglifyjs - > bundles/bundle.js",
     "browsertest": "yarn start-test; browserify test/client/*.js test/*.js | uglifyjs - | tape-run; yarn stop-test",
-    "browsertest-client": "yarn start-test; browserify test/client/*.js | uglifyjs - | tape-run --browser chrome; yarn stop-test",
```
Comment (diracdeltas) [inline package.json]: was this change intentional?

File: client/constants/messages.js [hunk 6]
```diff
@@ -45,12 +45,24 @@ const messages = {
...
+   * webview resolves sync records against matching browser objects and returns
+   * resolved syncRecords to apply against browser data.
+   * param looks like [[<syncRecord>, <browserObject=>], ...]
+   */
+  RESOLVE_SYNC_RECORDS: _, /* @param {Array.<Array.<Object>>} recordsAndExistingObjects */
+  /**
+   * webview -> browser
+   * browser must update its local values with the resolved sync records.
+   */
+  RESOLVED_SYNC_RECORDS: _, /* @param {Array.<Object>} records */
```
Comment (diracdeltas) [inline client/constants/messages.js]: given that this doesn't have a categoryName parameter, it seems the browser has to infer category from the record object anyway, so categoryName probably isn't needed for RECEIVE_SYNC_RECORDS either

File: client/constants/messages.js (hunk 4 above)
Comment (ayumi) [inline client/constants/messages.js]: sync's requestUtil still needs to make separate s3 ListObject requests per category, so in case a user syncs a subset of the options it would be a little faster.

we could not worry about that, but if it's already implemented i think it's ok to leave in categories for Receive?

File: package.json (hunk 5 above)
Comment (ayumi) [inline package.json]: yes; browsertest-client is helpful for developing client-only code because it runs faster. it's only used in dev, so it's ok to skip uglify as long as you check browsertest before committing.

File: client/constants/messages.js (hunk 4 above)
Comment (diracdeltas) [inline client/constants/messages.js]: i see, that means `RESOLVE_SYNC_RECORDS` and `RESOLVED_SYNC_RECORDS` will only ever send records in the same category, right? so it seems those should both send a categoryName parameter

Comment (ayumi) [inline client/constants/messages.js]: correct, those two calls would only call with same-category records.
browser clients can examine each record's objectData to determine what to do with them.

Comment (diracdeltas) [inline client/constants/messages.js]: up to you but i think it is slightly neater to send categoryName explicitly because all records will be in the same category
